[ecobee] Fix for error code 14 (token expired)

mhilbush opened the pull request:
The error code 14 (token expired) returned by the Ecobee API is not handled correctly by the binding. No matter how much time passes, the AccessTokenResponse `isExpired` method never returns true. I don't know why this is the case. I can only assume there's something wrong with the access token returned from Ecobee.

Therefore, the only way to recover from this condition (short of the user manually disabling and enabling the account thing) is to close and recreate the OAuthClientService (which is essentially what happens when you disable/enable the account thing).

Also raising the log level from DEBUG to INFO to give more visibility in the log to when the condition occurs.

mhilbush commented:
It also would be great if this could be cherry picked to 4.1.x

jlaur commented:
@mhilbush - a few comments about things that could be checked (if you didn't already) before introducing this work-around. First, I'm wondering why it registers itself as a listener:
https://github.com/openhab/openhab-addons/blob/c4f166dcdc8dc5e35d4ce1877de0317fbafaae88/bundles/org.openhab.binding.ecobee/src/main/java/org/openhab/binding/ecobee/internal/api/EcobeeApi.java#L127

but without actually implementing it:
https://github.com/openhab/openhab-addons/blob/c4f166dcdc8dc5e35d4ce1877de0317fbafaae88/bundles/org.openhab.binding.ecobee/src/main/java/org/openhab/binding/ecobee/internal/api/EcobeeApi.java#L205-L207

The method should probably set the instance variable `accessTokenResponse`, but in the current situation it might be even more useful to log something here to gain information about things happening before the token is apparently suddenly expired and during the renewal.

In the end, I think it should be decided if you'd like to subscribe to those updates, or update by hand in the binding code as needed (which is currently implemented).

In this place:
https://github.com/openhab/openhab-addons/blob/c4f166dcdc8dc5e35d4ce1877de0317fbafaae88/bundles/org.openhab.binding.ecobee/src/main/java/org/openhab/binding/ecobee/internal/api/EcobeeApi.java#L155-L171

you could also consider adding some more logging temporarily to see contents of `AccessTokenResponse`, especially `createdOn` and `expiresIn`.

I assume the clock is correctly set on the server running openHAB? (Only asking because that's a crucial part of the calculation)

mhilbush commented:
Thanks for the feedback. I'm trying to remember a few things. Here goes...

> First, I'm wondering why it registers itself as a listener: but without actually implementing it:

If I recall correctly, I had it this way initially. But the default `BUFFER_EXPIRES_IN_TOKEN_SECONDS` wasn't enough time for the Ecobee OAuth implementation. And, I recall there wasn't an API exposed for changing the default. So I had to do the refresh check manually. I guess I never removed the listener stuff.

> but in the current situation it might be even more useful to log something here to gain information about things happening before the token is apparently suddenly expired and during the renewal.

I agree it would be good to add some debug about the AccessTokenResponse. I didn't want to do it every time `isAuthorized` is called, so maybe I'll do it only when I get the error code 14.

> In the end, I think it should be decided if you'd like to subscribe to those updates, or update by hand in the binding code as needed (which is currently implemented).

For the reason mentioned above, I think I should leave the manual refresh process in place. And, therefore remove the listener stuff.

> you could also consider adding some more logging temporarily to see contents of AccessTokenResponse, especially createdOn and expiresIn.

Agreed. I'll do this. Maybe it will reveal some clues next time the problem happens.

> I assume the clock is correctly set on the server running openHAB? (Only asking because that's a crucial part of the calculation)

Yes, the clock is set correctly. And, this happens to other users, as well. Since the problems all seem to occur at the same time for all of us, I inferred that it's a temporary Ecobee issue that triggers it.

Please let me know what you think about the above.

In the meantime, I'll do 2 things.
- remove the listener stuff since it's not used; and
- add some debug to log some access token response information when the error 14 occurs

lsiepel commented on bundles/org.openhab.binding.ecobee/src/main/java/org/openhab/binding/ecobee/internal/api/EcobeeApi.java:
```diff
@@ -335,7 +337,7 @@ private boolean isSuccess(@Nullable AbstractResponseDTO response) {
             logger.debug("API: Ecobee API returned null response");
         } else if (response.status.code.intValue() != 0) {
-            logger.debug("API: Ecobee API returned unsuccessful status: code={}, message={}", response.status.code,
+            logger.info("API: Ecobee API returned unsuccessful status: code={}, message={}", response.status.code,
```
Please revert to debug.

mhilbush replied:
Done

lsiepel commented on bundles/org.openhab.binding.ecobee/src/main/java/org/openhab/binding/ecobee/internal/api/EcobeeApi.java:
```diff
@@ -344,11 +346,13 @@ private boolean isSuccess(@Nullable AbstractResponseDTO response) {
             } else if (response.status.code == ECOBEE_TOKEN_EXPIRED) {
                 // Check isAuthorized again to see if we can get a valid token
-                logger.debug("API: Unable to complete API call because token is expired");
+                logger.info("API: Unable to complete API call because token is expired. Try to refresh the token...");
```
Please revert to debug.

mhilbush replied:
Done

lsiepel commented on bundles/org.openhab.binding.ecobee/src/main/java/org/openhab/binding/ecobee/internal/api/EcobeeApi.java:
```diff
@@ -344,11 +346,13 @@ private boolean isSuccess(@Nullable AbstractResponseDTO response) {
                     return true;
                 } else {
-                    logger.debug("API: isAuthorized was NOT successful on second try");
+                    logger.info("API: isAuthorized was NOT successful on second try");
```
I guess this is a final state? If so this should be warn or the Thing status should be set (i'm unable to determine this in this short review)

mhilbush replied:
Yes, you're probably right that the status should be set to OFFLINE.

lsiepel reviewed (changes requested):
Thanks, Left some comments.

Edit: Oh sorry, i totally missed @jlaur 's comments. This review seems outdated.

jlaur commented:
> > you could also consider adding some more logging temporarily to see contents of AccessTokenResponse, especially createdOn and expiresIn.
> 
> Agreed. I'll do this. Maybe it will reveal some clues next time the problem happens.

👍

> > I assume the clock is correctly set on the server running openHAB? (Only asking because that's a crucial part of the calculation)
> 
> Yes, the clock is set correctly. And, this happens to other users, as well. Since the problems all seem to occur at the same time for all of us, I inferred that it's a temporary Ecobee issue that triggers it.

Roger that. The only possibility that comes to my mind how that could happen is if `expires_in` in the [access token response](https://datatracker.ietf.org/doc/html/rfc6749#section-4.2.2) from the server is longer than the actual validity time. `createdOn` is updated locally by openHAB.

With added debug logs you should be able to see the value of `expiresIn` right after getting a new token, and if you have both `createdOn` and `expiresIn` when getting error code 14, you should be able to verify if in fact this happens while the token should still be valid according to those values.

jlaur commented on bundles/org.openhab.binding.ecobee/src/main/java/org/openhab/binding/ecobee/internal/api/EcobeeApi.java:
```diff
@@ -344,11 +346,13 @@ private boolean isSuccess(@Nullable AbstractResponseDTO response) {
                 // Check isAuthorized again to see if we can get a valid token
-                logger.debug("API: Unable to complete API call because token is expired");
+                logger.info("API: Unable to complete API call because token is expired. Try to refresh the token...");
+                closeOAuthClientService();
+                createOAuthClientService();
```
The other comments aside, assuming that a work-around will in fact be needed, do you think this would be enough (and more light-weight)?
```suggestion
                oAuthClientService.refreshToken()
```

mhilbush replied:
Yes, good idea. I think that will work.

jlaur replied:
I think you will then need remove lines 356-358 to not do both. 😉 Additionally, after triggering the token refresh, you should probably:
```java
return isAuthorized();
```

or similar. `refreshToken()` returns the new token, but `isAuthorized()` will store it, validate it and use it, I think.

mhilbush commented:
> The only possibility that comes to my mind how that could happen is if expires_in in the [access token response](https://datatracker.ietf.org/doc/html/rfc6749#section-4.2.2) from the server is longer than the actual validity time. createdOn is updated locally by openHAB.
> 
>With added debug logs you should be able to see the value of expiresIn right after getting a new token, and if you have both createdOn and expiresIn when getting error code 14, you should be able to verify if in fact this happens while the token should still be valid according to those values.

Ok, good to know. So if I log `createdOn` and `expiresIn`, it should give me a clue. I looked at `isExpired`. Other than what I pass in (current time and expiresInBufferSeconds), it only uses `createdOn` and `expiresIn`.

It might be a while before I get anything in the log. Th eproblem only occurs very infrequently, sometimes going quite a few months without the problem occurring.

mhilbush commented:
@jlaur @lsiepel I believe I addressed all your review feedback.

@jlaur However...

> The only possibility that comes to my mind how that could happen is if expires_in in the [access token response](https://datatracker.ietf.org/doc/html/rfc6749#section-4.2.2) from the server is longer than the actual validity time. 

Now that I think about this... If what you say is true - that expires_in is no longer the actual validity time, rather than check `isExpired` like I was doing, maybe I should've just called `refreshToken` and not bother to see if the token was expired.

Here's what I would propose. Since I know my workaround resolves the issue, I think I would prefer to have the PR merged as is. Then, I'll make a version of the binding that just calls refreshToken and run that in my environment until the problem occurs again (possibly a few months from now). And if it works, I'll do another PR to remove my workaround. WDYT?

jlaur commented:
> Now that I think about this... If what you say is true - that expires_in is no longer the actual validity time, rather than check `isExpired` like I was doing, maybe I should've just called `refreshToken` and not bother to see if the token was expired.

I think you can still check `isExpired` for RFC compliance, i.e. if true, trigger a token refresh right away (like before your changes). There is no cost for doing that. In addition to that, you can then refresh the token also when receiving error code 14. I'm not sure if this is what you mean? If you mean always unconditionally refresh the token in `isAuthorized()`, then I'm not sure it's a good idea, because you would have many unneeded token refreshes.

> Here's what I would propose. Since I know my workaround resolves the issue, I think I would prefer to have the PR merged as is. Then, I'll make a version of the binding that just calls refreshToken and run that in my environment until the problem occurs again (possibly a few months from now). And if it works, I'll do another PR to remove my workaround. WDYT?

Sure, we can definitely merge something now without waiting for additional logs. We just need to agree on the fix we'll be merging, according to my comments above.

mhilbush commented:
I put the call to `isAuthorized` back. This is what was working, and this is what I would prefer to go with if you're ok with it.

Then, I'll make the following change to replace the calls to recreate the OAuthClientService, and I'll run it in my prod environment until the problem occurs again. If this works, then I'll submit another PR with this change.

```
                // Try to refresh the token
                try {
                    oAuthClientService.refreshToken();
                    if (isAuthorized()) {
                        return true;
                    } else {
                        logger.warn("API: isAuthorized was NOT successful on second try");
                        bridgeHandler.updateBridgeStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
                                "Unable to refresh access token");
                    }
                } catch (OAuthException | IOException | OAuthResponseException e) {
                    logger.warn("API: Exception trying to refresh the access token", e);
                    bridgeHandler.updateBridgeStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
                            "Unable to refresh access token");
                }
```

jlaur commented:
> I put the call to `isAuthorized` back. This is what was working, and this is what I would prefer to go with if you're ok with it.
> 
> Then, I'll make the following change to replace the calls to recreate the OAuthClientService, and I'll run it in my prod environment until the problem occurs again. If this works, then I'll submit another PR with this change.

I've given it yet another look and some thoughts. Allow me to try to convince you for a few changes - sorry. 😉

First a new observation: `oAuthClientService.getAccessTokenResponse()` already checks if access token has expired, and if there is a refresh token, the access token will be automatically updated. This could very well be the reason that your expiration check is never fulfilled - simply because it's already renewed and thus not expired. The only difference is the `tokenExpiresInBuffer` parameter which is defaulted to 10 seconds, but provided as 120 seconds in your custom call.

So maybe the problem isn't that the token expires before it should, but could also be a network issue/failure when calling `getAccessTokenResponse()`. However, that would result in a warning being logged, so if this hasn't been observed, that's not it. Another hypothesis could be that an Ecobee server deployment/restart (or similar) forces expiration ahead of time, i.e. an issue on their side, which would be exactly what you are trying to mitigate.

That's just an observation, now let me propose something. First, we know that `isAuthorized()` is working, i.e. it's able to obtain and refresh the token most of the time. Therefore, this method could be slightly refactored with minimal risk:
- Add a new parameter: `boolean forceTokenRefresh`.
- The condition `localAccessTokenResponse.isExpired(Instant.now(), TOKEN_EXPIRES_IN_BUFFER_SECONDS)` is changed to `forceTokenRefresh || localAccessTokenResponse.isExpired(Instant.now(), TOKEN_EXPIRES_IN_BUFFER_SECONDS)`.
- In `isSuccess()` the call `isAuthorized()` is changed to `isAuthorized(true)`. Other calls are changed to `isAuthorized(false)`.
- This is not really clean code, only proposed as a quick and dirty way to minimize risk.

I realize the contradiction of coming up with a theory that the expiration check was never effective, and then proposing that reusing already working code is safe. Therefore, feel free to take the ideas you feel comfortable with and discard the rest.

Oh, and it should also be pretty safe to no longer implement `AccessTokenRefreshListener`, since it's not used - like previously discussed.

mhilbush commented:
> I've given it yet another look and some thoughts. Allow me to try to convince you for a few changes - sorry. 😉

No, it's good to talk this through. 😉

> First a new observation: oAuthClientService.getAccessTokenResponse() already checks if access token has expired, and if there is a refresh token, the access token will be automatically updated. This could very well be the reason that your expiration check is never fulfilled - simply because it's already renewed and thus not expired. The only difference is the tokenExpiresInBuffer parameter which is defaulted to 10 seconds, but provided as 120 seconds in your custom call.

But if it was already renewed, then subsequent API calls would succeed, wouldn't they. In fact, once the error code 14 occurs, all future API calls fail with error code 14 unless the binding or openHAB is restarted, or the account bridge thing is disabled/enabled, all of which cause a new OAuthClientService to be created.

> So maybe the problem isn't that the token expires before it should, but could also be a network issue/failure when calling getAccessTokenResponse(). However, that would result in a warning being logged, so if this hasn't been observed, that's not it.

Agreed. Also, it would have to be a network issue on the Ecobee side because when the problem occurs it impacts multiple users.

> Another hypothesis could be that an Ecobee server deployment/restart (or similar) forces expiration ahead of time, i.e. an issue on their side, which would be exactly what you are trying to mitigate.

Yes, agreed. This seems to be the most likely scenario.

> That's just an observation, now let me propose something. First, we know that isAuthorized() is working, i.e. it's able to obtain and refresh the token most of the time. Therefore, this method could be slightly refactored with minimal risk:

Ok, thanks, I'll take a look at this.

> I realize the contradiction of coming up with a theory that the expiration check was never effective, and then proposing that reusing already working code is safe. Therefore, feel free to take the ideas you feel comfortable with and discard the rest.

Thanks, I appreciate that.

> Oh, and it should also be pretty safe to no longer implement AccessTokenRefreshListener, since it's not used - like previously discussed.

Yes, of course. Sorry.

mhilbush commented:
Removed AccessTokenRefreshListener implementation.

Still looking at your suggested refactoring of isAuthorized.

Also related. Completely lost track of this. 🙄 🙀
https://github.com/openhab/openhab-core/issues/1438

lsiepel commented on bundles/org.openhab.binding.ecobee/src/main/java/org/openhab/binding/ecobee/internal/api/EcobeeApi.java:
```diff
@@ -337,25 +338,34 @@ private boolean isSuccess(@Nullable AbstractResponseDTO response) {
+                } catch (OAuthException | IOException | OAuthResponseException e) {
+                    logger.warn("API: Unable to refresh token after receiving error code 14");
+                    bridgeHandler.updateBridgeStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                            "Unable to refresh access token");
```
```suggestion
                            "Unable to refresh access token, code 14");
```

If add sdomehting like this to the thing state, the logger can be removed as it does not add anything to the logging that is allready there.

mhilbush replied:
@lsiepel I'm a little confused. If you're logging debug to openhab.log and logging thing ONLINE/OFFLINE status changes to events.log, isn't it a hassle to be looking across 2 log files to get the full picture of what's going on? I must be missing something.

lsiepel replied:
It is a general rule to only log an issue once. Last bit of: https://www.openhab.org/docs/developer/guidelines.html#f-logging

mhilbush replied:
I understand it's a general rule. And I agree there are many, many cases where it doesn't make sense, such as the rather obvious example given in the docs. I just don't understand why you wouldn't want to se an error condition in the log file, especially when the error condition in this context has been shown to occur only every few months. And in this scenario where there's other relevant information being logged to debug.

lsiepel replied:
I wasn't aware of this low frequency.

Seen some log viewers that combine both logs. If you manually check the logs seperate I can see your use case. Not going to maker an issue of it.

mhilbush replied:
Thanks. Yeah, a coule things have to go wrong here. First, you get the error code 14 from Ecobee which happens very rarely (every few months), then you get an exception trying to refresh the access token after having just successfully created the OAuthClientService.

lsiepel commented:
Do you still need to look at the refactoring of isAuthorized or is this ready to be merged?

mhilbush commented:
I wanted to look at the refactoring first. Thanks

mhilbush commented:
@jlaur 

I like your suggestion to slightly refactor isAuthorized; however, that presupposes that refreshing the access token after an error code 14 will work. Since I haven't established that yet, I would prefer to merge this PR as is since I've proven that this resolves the issue (assuming you are ok with the current state of the change).

In the meantime, I'm running a version that refactors isAuthorized to force a call to refreshToken after an error code 14. Next time the error code 14 occurs, and assuming the binding recovers successfully, I'll submit this as a separate PR.

Are you ok with this approach?

jlaur commented:
> @jlaur
> 
> I like your suggestion to slightly refactor isAuthorized; however, that presupposes that refreshing the access token after an error code 14 will work. Since I haven't established that yet, I would prefer to merge this PR as is since I've proven that this resolves the issue (assuming you are ok with the current state of the change).
> 
> In the meantime, I'm running a version that refactors isAuthorized to force a call to refreshToken after an error code 14. Next time the error code 14 occurs, and assuming the binding recovers successfully, I'll submit this as a separate PR.
> 
> Are you ok with this approach?

That sounds sensible. Let's give this first version of the fix some mileage, and hope the issue will reoccur on your system soon. 🙂

jlaur reviewed (approved):
LGTM

mhilbush commented:
Thanks, @jlaur 

And thanks for cherry picking this for the next 4.1 patch release.

mhilbush commented:
FWIW, I got what looks like 6 or 7 error code 14s in a row this afternoon at around 1:55 pm US EST. I use logreader binding to detect and send me emails when that error occurs.

When I get home I will look in the logs to see what happened. It looks like the binding recovered from it but I won’t know for sure until I see the logs. Hopefully I got some good debug info.

mhilbush commented:
Well, this is interesting. And it isn't exactly what I expected. As a reminder, I'm running a special test build of the binding that forces a token refresh when error code 14 occurs, rather than close/create the OAuthClientService.

```
2024-03-07 13:54:14.586 [DEBUG] [org.openhab.binding.ecobee.internal.api.EcobeeApi ] - API: Ecobee API returned unsuccessful status: code=14, message=Authentication token has expired. Refresh your tokens.
2024-03-07 13:54:14.586 [DEBUG] [org.openhab.binding.ecobee.internal.api.EcobeeApi ] - API: Unable to complete API call because token is expired. Try to refresh the token...
2024-03-07 13:54:14.586 [DEBUG] [org.openhab.binding.ecobee.internal.api.EcobeeApi ] - API: AccessTokenResponse created on: 2024-03-07T18:04:02.943894652Z
2024-03-07 13:54:14.586 [DEBUG] [org.openhab.binding.ecobee.internal.api.EcobeeApi ] - API: AccessTokenResponse expires in: 3600
```

Token was created at 13:04:02 (US EST) with 1 hour expiration. Yet at 13:54:14, Ecobee said the token was expired. This clearly points to something wrong on their end. Possibly as you said @jlaur, they restarted a server(s).

The binding then tries to force a token refresh, which fails miserably.

```
2024-03-07 13:54:14.586 [DEBUG] [org.openhab.binding.ecobee.internal.api.EcobeeApi ] - API: Token is expiring soon. Refresh it now
2024-03-07 13:54:15.480 [DEBUG] [org.openhab.binding.ecobee.internal.api.EcobeeApi ] - API: Perform query on thermostat: '[311078994752, 413729395643]'
2024-03-07 13:54:15.535 [DEBUG] [org.openhab.binding.ecobee.internal.api.EcobeeApi ] - API: Ecobee API returned unsuccessful status: code=14, message=Authentication token has expired. Refresh your tokens.
2024-03-07 13:54:15.535 [DEBUG] [org.openhab.binding.ecobee.internal.api.EcobeeApi ] - API: Unable to complete API call because token is expired. Try to refresh the token...
2024-03-07 13:54:15.535 [DEBUG] [org.openhab.binding.ecobee.internal.api.EcobeeApi ] - API: AccessTokenResponse created on: 2024-03-07T18:54:15.476384532Z
2024-03-07 13:54:15.535 [DEBUG] [org.openhab.binding.ecobee.internal.api.EcobeeApi ] - API: AccessTokenResponse expires in: 3600
```

Here Ecobee still thinks the token is expired even though it was just refreshed. Bad Ecobee.

So we force another token refresh, which ends even more badly lol.

```
2024-03-07 13:54:15.536 [DEBUG] [org.openhab.binding.ecobee.internal.api.EcobeeApi ] - API: Token is expiring soon. Refresh it now
2024-03-07 13:54:15.861 [WARN ] [org.openhab.binding.ecobee.internal.api.EcobeeApi ] - API: Got exception trying to get access token from OAuth service
org.openhab.core.auth.client.oauth2.OAuthException: Exception in oauth communication, grant type refresh_token: Bad http response, http code 429
        at org.openhab.core.auth.oauth2client.internal.OAuthConnector.doRequest(OAuthConnector.java:352) ~[?:?]
        at org.openhab.core.auth.oauth2client.internal.OAuthConnector.grantTypeRefreshToken(OAuthConnector.java:197) ~[?:?]
        at org.openhab.core.auth.oauth2client.internal.OAuthClientServiceImpl.refreshToken(OAuthClientServiceImpl.java:316) ~[?:?]
        at org.openhab.binding.ecobee.internal.api.EcobeeApi.isAuthorized(EcobeeApi.java:165) ~[?:?]
        at org.openhab.binding.ecobee.internal.api.EcobeeApi.isSuccess(EcobeeApi.java:357) ~[?:?]
        at org.openhab.binding.ecobee.internal.api.EcobeeApi.performThermostatQuery(EcobeeApi.java:249) ~[?:?]
        at org.openhab.binding.ecobee.internal.handler.EcobeeAccountBridgeHandler.refreshThermostats(EcobeeAccountBridgeHandler.java:223) ~[?:?]
        at org.openhab.binding.ecobee.internal.handler.EcobeeAccountBridgeHandler.refresh(EcobeeAccountBridgeHandler.java:214) ~[?:?]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539) [?:?]
        at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:305) [?:?]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:305) [?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) [?:?]
        at java.lang.Thread.run(Thread.java:840) [?:?]
Caused by: org.openhab.core.auth.client.oauth2.OAuthException: Bad http response, http code 429
        at org.openhab.core.auth.oauth2client.internal.OAuthConnector.doRequest(OAuthConnector.java:342) ~[?:?]
        ... 13 more
2024-03-07 13:54:15.863 [WARN ] [org.openhab.binding.ecobee.internal.api.EcobeeApi ] - API: isAuthorized was NOT successful forcing the access token refresh
2024-03-07 13:54:46.870 [DEBUG] [org.openhab.binding.ecobee.internal.api.EcobeeApi ] - API: Perform thermostat summary query
2024-03-07 13:54:46.921 [DEBUG] [org.openhab.binding.ecobee.internal.api.EcobeeApi ] - API: Ecobee API returned unsuccessful status: code=14, message=Authentication token has expired. Refresh your tokens.
2024-03-07 13:54:46.921 [DEBUG] [org.openhab.binding.ecobee.internal.api.EcobeeApi ] - API: Unable to complete API call because token is expired. Try to refresh the token...
```

After a few cycles of this, the binding eventually recovers.

This tells me that the close/create of the OAuthClientService is the better workaround, since it recovers immediately, rather than failing several times before recovering.

I'll continue running my test version to see what happens when this happens again.
